src/main/java/io/confluent/idesidecar/restapi/kafkarest/SchemaManager.java

```diff
@@ -0,0 +1,166 @@
+              + "schema ID, subject, subject name strategy, type, or schema."
+      );
+    }
+
+    return null;
+  }
+
+  /**
+   * Check if the ProduceRequestData contains a non-null schemaVersion and all other
+   * schema related fields are null.
+   */
+  private static boolean supportsSchemaVersion(ProduceRequestData produceRequestData) {
+    // Only schemaVersion must be set
+    return produceRequestData.getSchemaVersion() != null
+        && !unsupportedFieldsSet(produceRequestData);
+  }
+
+  @SuppressWarnings("BooleanExpressionComplexity")
+  private static boolean unsupportedFieldsSet(ProduceRequestData produceRequestData) {
+    return (produceRequestData.getSchemaId() != null
+        || produceRequestData.getSubject() != null)
+        || produceRequestData.getSubjectNameStrategy() != null
+        || produceRequestData.getType() != null
+        || produceRequestData.getSchema() != null;
+  }
+
+  private RegisteredSchema getSchemaFromSchemaVersion(
+      @Nullable SchemaRegistryClient schemaRegistryClient,
+      String topicName,
+      Integer schemaVersion,
+      boolean isKey
+  ) {
+    if (schemaRegistryClient == null) {
+      if (schemaVersion != null) {
+        throw new BadRequestException("Schema version requested without a schema registry client");
+      }
+      return null;
+    }
+
+    var schema = schemaRegistryClient.getByVersion(
+        // Note: We default to TopicNameStrategy for the subject name for the sake of simplicity.
+        (isKey ? topicName + "-key" : topicName + "-value"),
+        schemaVersion,
+        // do not lookup deleted schemas
+        false
+    );
+    return new RegisteredSchema(
+        schema.getSubject(), schema.getId(), schema.getVersion(), parseSchema(schema)
+    );
+  }
+
+  private ParsedSchema parseSchema(Schema schema) {
+    var schemaFormat = SchemaFormat.fromSchemaType(schema.getSchemaType());
+    var schemaProvider = Optional
+        .ofNullable(schemaFormat.schemaProvider())
+        .orElseThrow(() ->
+            new IllegalArgumentException("Schema type has no provider: " + schema.getSchemaType()));
+    return schemaProvider
+        .parseSchema(schema, false)
+        .orElseThrow(() -> new BadRequestException("Failed to parse schema"));
+  }
+
+  public enum SchemaFormat {
+    AVRO {
+      private final SchemaProvider schemaProvider = new AvroSchemaProvider();
+
+      @Override
+      SchemaProvider schemaProvider() {
+        return schemaProvider;
+      }
+    },
+    PROTOBUF {
+      private final SchemaProvider schemaProvider = new ProtobufSchemaProvider();
+
+      @Override
+      SchemaProvider schemaProvider() {
+        return schemaProvider;
+      }
+    },
+    JSON {
+      private final SchemaProvider schemaProvider = new JsonSchemaProvider();
+
+      @Override
+      SchemaProvider schemaProvider() {
+        return schemaProvider;
+      }
+    };
+
+    abstract SchemaProvider schemaProvider();
```

Comment:
This can be simplified a bit:
```suggestion
  public enum SchemaFormat {
    AVRO(new AvroSchemaProvider()),
    PROTOBUF(new ProtobufSchemaProvider()),
    JSON(new JsonSchemaProvider());
    
    private final SchemaProvider schemaProvider;

    SchemaFormat(SchemaProvider schemaProvider) {
      this.schemaProvider = schemaProvider;
    }

    SchemaProvider schemaProvider() {
      return schemaProvider;
    }
```

Reply:
Oh, yeah that's much better, thanks!